internal/httputil/reproxy/reproxy.go

```diff
@@ -0,0 +1,144 @@
+// Package reproxy contains a handler for re-proxying traffic through the http controlplane.
+package reproxy
+
+import (
+	"crypto/cipher"
+	"encoding/base64"
+	"math/rand"
+	"net/http"
+	stdhttputil "net/http/httputil"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/pomerium/pomerium/config"
+	"github.com/pomerium/pomerium/internal/httputil"
+	"github.com/pomerium/pomerium/internal/log"
+	"github.com/pomerium/pomerium/pkg/cryptutil"
+)
+
+// The Handler looks for an X-Pomerium-Reproxy-Policy header and if found re-proxies the request upstream
+// to the destination.
+//
+// This is used to forward requests to Kubernetes with headers split to multiple values instead of coalesced via a
+// comma.
+type Handler struct {
+	mu       sync.RWMutex
+	cipher   cipher.AEAD
+	options  *config.Options
+	policies map[uint64]*config.Policy
+}
+
+// New creates a new Handler.
+func New() *Handler {
+	h := new(Handler)
+	h.cipher, _ = cryptutil.NewAEADCipher(cryptutil.NewKey())
```

Comment:
I think this can only fail if we give it the wrong key size or run out of entropy.

Reply:
Could there be any issues by generating a new key for each for each instance? Presumably a request will only be re...proxied to itself but I want to be sure.

Reply:
The shared key is used.